### PR TITLE
Test metrics

### PR DESCRIPTION
**Pull request by smguo**
This PR added computing test metrics and saving as a csv file during model inference. Sorry for the messy commit history... I'll rebase instead of merge changes from master next time

### CODE REVIEW

**Comment by jennyfolkesson** [comment]
@smguo thanks! Tests are failing here too, you can see build history here to fix them:
https://travis-ci.com/czbiohub/microDL/builds/103469306

**Comment by smguo** [comment]
Thanks @jennyfolkesson! Tests are fixed now.

File: micro_dl/cli/image_inference.py [hunk 1]
```diff
@@ -106,8 +115,25 @@ def run_prediction(args, gpu_ids, gpu_mem_frac):
...
+                              index_col=0)
+    test_tile_meta = pd.read_csv(os.path.join(args.model_dir, 'test_metadata.csv'),
+                              index_col=0)
+    # TODO: generate test_frames_meta.csv together with tile csv during training
+    test_frames_meta_filename = os.path.join(args.model_dir, 'test_frames_meta.csv')
+    metrics = trainer_config['metrics']
+    if args.metrics:
+        metrics = args.metrics
+    if isinstance(metrics, str):
+        metrics = [metircs]
```
Comment (anithapk) [inline micro_dl/cli/image_inference.py]: spelling metircs -> metrics

File: micro_dl/cli/image_inference.py [hunk 2]
```diff
@@ -106,8 +115,25 @@ def run_prediction(args, gpu_ids, gpu_mem_frac):
...
         config = yaml.load(f)
     # Load frames metadata and determine indices
-    frames_meta = pd.read_csv(os.path.join(args.image_dir, 'frames_meta.csv'))
-    split_idx_name = config['dataset']['split_by_column']
+    network_config = config['network']
+    dataset_config = config['dataset']
+    trainer_config = config['trainer']
+    frames_meta = pd.read_csv(os.path.join(args.image_dir, 'frames_meta.csv'),
+                              index_col=0)
+    test_tile_meta = pd.read_csv(os.path.join(args.model_dir, 'test_metadata.csv'),
```
Comment (anithapk) [inline micro_dl/cli/image_inference.py]: sorry, test_metadata is frames_meta['pos_idx'] in split_idx['test']. Is this different than what we had before? Looks like both approaches will get the same output! what am i missing?

File: micro_dl/cli/image_inference.py [hunk 3]
```diff
@@ -229,33 +261,64 @@ def run_prediction(args, gpu_ids, gpu_mem_frac):
...
+                im_target = preprocess_imstack(
+                    frames_metadata=frames_meta,
+                    input_dir=args.image_dir,
+                    depth=1,
+                    time_idx=time_idx,
+                    channel_idx=target_channel,
+                    slice_idx=slice_idx,
+                    pos_idx=pos_idx,
+                )
+                im_target = image_utils.crop2base(im_target)
```
Comment (anithapk) [inline micro_dl/cli/image_inference.py]: this will fail with the current crop2base as crop2base expects a 2D image and crops dim=[0, 1] only. Here after preprocess_imstack, u've to crop along dim=[1, 2] for stack to 2D! u can get the changes from 3D_inference: https://github.com/czbiohub/microDL/blob/b3b8f20a079829b7874238dcbf45c2af8cf37756/micro_dl/utils/image_utils.py#L77, however i haven't tested it yet!

File: micro_dl/cli/image_inference.py [hunk 4]
```diff
@@ -229,33 +261,64 @@ def run_prediction(args, gpu_ids, gpu_mem_frac):
...
+                if data_format == 'channels_first':
```
Comment (anithapk) [inline micro_dl/cli/image_inference.py]: preprocess_imstack adds new axis at axis=2, by default assembles in xyz format, the check here should be on image_format I believe?

File: micro_dl/cli/image_inference.py [hunk 5]
```diff
@@ -229,33 +261,64 @@ def run_prediction(args, gpu_ids, gpu_mem_frac):
...
+                    # Change dimension order to zyx (3D) or channel first (2D)
+                    im_target = np.transpose(im_target, [2, 0, 1])
+                im_target = im_target[np.newaxis, ...]
```
Comment (anithapk) [inline micro_dl/cli/image_inference.py]: this is the case for data_format == 'channels_first'

File: micro_dl/cli/image_inference.py [hunk 6]
```diff
@@ -229,33 +261,64 @@ def run_prediction(args, gpu_ids, gpu_mem_frac):
...
+                if len(im_target.shape) < len(im_stack.shape):
```
Comment (anithapk) [inline micro_dl/cli/image_inference.py]: this should happen only for StackTo2D as we go from 5D tensor to 4D tensor!

File: micro_dl/config_preprocess_kidney_confocal.yml [hunk 7]
```diff
@@ -4,16 +4,18 @@ input_dir: '/data/sguo/Processed/2018_11_01_kidney_slice/SMS_2018_1101_1906_1_BG
...
     str_elem_radius: 3
+do_tiling: True
 tile:
     channels: -1
     tile_size: [256, 256]
     step_size: [128, 128]
     depths: [3, 3, 3, 3, 1, 1, 1]
     mask_depth: 1
-    image_format: 'zyx'
+    shape_order: 'zyx'
```
Comment (anithapk) [inline micro_dl/config_preprocess_kidney_confocal.yml]: i think we changed this to image_format!

File: micro_dl/input/training_table.py [hunk 8]
```diff
@@ -86,7 +86,7 @@ def train_test_split(self):
             "into train/val".format(self.split_by_column)
         # DOES NOT HANDLE NON-INTEGER VALUES. map to int if string
         # the sample_idxs are required for evaluating performance on test set
-
+        unique_values = unique_values.astype(np.uint32)
```
Comment (anithapk) [inline micro_dl/input/training_table.py]: wondering what is the rationale for this? this would make the assert below unnecessary. If unique_values is not an integer (say label_names), it will not raise an exception!

File: micro_dl/train/metrics.py [hunk 9]
```diff
@@ -82,3 +81,15 @@ def ssim(y_true, y_pred, max_val=6):
...
+    """Pearson correlation
+    :param tensor y_true: Labeled ground truth
+    :param tensor y_pred: Predicted label,
+    :param float max_val: The dynamic range of the images (i.e., the
+        difference between the maximum the and minimum allowed values).
+    :return float : Pearson over all images in the batch
+    """
+    covariance = K.mean((y_pred - K.mean(y_pred))
+                        * (y_true - K.mean(y_true)))
+    return covariance/K.std(y_pred)/K.std(y_true)
```
Comment (anithapk) [inline micro_dl/train/metrics.py]: 💯

File: micro_dl/cli/image_inference.py [hunk 10]
```diff
@@ -122,10 +148,15 @@ def run_prediction(args, gpu_ids, gpu_mem_frac):
...
     for id in iter_ids:
         if id != split_idx_name:
-            metadata_ids[id] = np.unique(frames_meta[id])
-
+            metadata_ids[id] = np.unique(test_tile_meta[id])
+            n_test_row *= len(metadata_ids[id])
+    # create empty dataframe for test image metadata
+    test_frames_meta = pd.DataFrame(columns=
+                                    frames_meta.columns.values.tolist()+metrics,
+                                    index=range(n_test_row))
```
Comment (anithapk) [inline micro_dl/cli/image_inference.py]: 🥇

File: micro_dl/cli/image_inference.py (hunk 2 above)
Comment (anithapk) [reply micro_dl/cli/image_inference.py]: got it, to include only the slices used in tiling. :-)

File: micro_dl/input/training_table.py (hunk 8 above)
Comment (anithapk) [reply micro_dl/input/training_table.py]: does pandas unique() not return an int array? I never checked that...

File: micro_dl/cli/image_inference.py [hunk 11]
```diff
@@ -169,7 +200,8 @@ def run_prediction(args, gpu_ids, gpu_mem_frac):
         predict=True,
     )
     print(model.summary())
-
+    model.compile(loss=loss_cls, optimizer='Adam', metrics=metrics_cls)
```
Comment (jennyfolkesson) [inline micro_dl/cli/image_inference.py]: I guess it doesn't matter much since we're not optimizing here, but you could get the optimizer from the config.

File: micro_dl/cli/image_inference.py (hunk 4 above)
Comment (jennyfolkesson) [reply micro_dl/cli/image_inference.py]: This is my bad, from when I used data and image format interchangeably. Now it should be if image_format == 'zyx'.

File: micro_dl/plotting/plot_utils.py [hunk 12]
```diff
@@ -32,12 +32,12 @@ def save_predicted_images(input_batch, target_batch, pred_batch,
     batch_size = len(input_batch)
     if batch_size == 1:
         assert output_fname is not None, 'need fname for saving image'
-        fname = os.path.join(output_dir, '{}.jpg'.format(output_fname))
+        fname = os.path.join(output_dir, '{}.{}'.format(output_fname, ext))
```
Comment (jennyfolkesson) [inline micro_dl/plotting/plot_utils.py]: 👍

File: micro_dl/cli/image_inference.py [hunk 13]
```diff
@@ -229,33 +261,64 @@ def run_prediction(args, gpu_ids, gpu_mem_frac):
...
+                    target_channel,
+                    slice_idx,
+                    pos_idx,
+                )
+                file_path = os.path.join(
+                    args.image_dir,
+                    frames_meta.loc[meta_idx, "file_name"],
+                )
+                # get a single row of frame meta data
+                test_frames_meta_row = frames_meta.loc[[meta_idx]]
```
Comment (jennyfolkesson) [inline micro_dl/cli/image_inference.py]: It might be cleaner to use .iloc[meta_idx]

File: micro_dl/cli/image_inference.py [hunk 14]
```diff
@@ -229,33 +261,64 @@ def run_prediction(args, gpu_ids, gpu_mem_frac):
                     np.save(file_name, im_pred, allow_pickle=True)
                 else:
                     raise ValueError('Unsupported file extension')
+
+                # assuming target and predicted images are always 2D for now
+                # Load target
+                meta_idx = aux_utils.get_meta_idx(
+                    frames_meta,
+                    time_idx,
+                    target_channel,
```
Comment (jennyfolkesson) [inline micro_dl/cli/image_inference.py]: It looks like target_channel is only declared if using the --save_figs flag right now. Since it's being used no matter what, it probably makes sense to move it outside of that if statement.

File: micro_dl/cli/image_inference.py (hunk 1 above)
Comment (smguo) [reply micro_dl/cli/image_inference.py]: Oops. Thanks for catching this!

File: micro_dl/config_preprocess_kidney_confocal.yml (hunk 7 above)
Comment (smguo) [reply micro_dl/config_preprocess_kidney_confocal.yml]: I'll delete this old config file as we don't need it in the repo

File: micro_dl/cli/image_inference.py (hunk 2 above)
Comment (smguo) [reply micro_dl/cli/image_inference.py]: Yes! Otherwise all the slices will be used for testing :-)

File: micro_dl/cli/image_inference.py (hunk 4 above)
Comment (smguo) [reply micro_dl/cli/image_inference.py]: I saw that the preprocessing config switched to 'image_format', 'data_format' is still used in the network config. Here the input image format has to match the the network input format. Should we use 'image_format' for both preprocessing and network configs?

File: micro_dl/cli/image_inference.py (hunk 5 above)
Comment (smguo) [reply micro_dl/cli/image_inference.py]: Yes, and line 293 is to add the batch dimension

File: micro_dl/input/training_table.py (hunk 8 above)
Comment (smguo) [reply micro_dl/input/training_table.py]: At some point I was getting indices in float type when I was debugging so I added this as a temporary fix... If the input are all integers pandas unique() should return integers. I'll remove it!

**Comment by smguo** [comment]
@jennyfolkesson Thanks for making the script work for 2.5D!

File: micro_dl/cli/image_inference.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (smguo) [reply micro_dl/cli/image_inference.py]: Thanks for catching this!

File: micro_dl/cli/image_inference.py [hunk 15]
```diff
@@ -106,8 +115,25 @@ def run_prediction(args, gpu_ids, gpu_mem_frac):
...
+                              index_col=0)
+    # TODO: generate test_frames_meta.csv together with tile csv during training
+    test_frames_meta_filename = os.path.join(args.model_dir, 'test_frames_meta.csv')
+    metrics = trainer_config['metrics']
+    if args.metrics:
+        metrics = args.metrics
+    if isinstance(metrics, str):
+        metrics = [metircs]
+    loss = trainer_config['loss']
+    metrics_cls = train_utils.get_metrics(args.metrics)
```
[the remark this answers is not in the document]
Comment (smguo) [reply micro_dl/cli/image_inference.py]: My bad! Thanks for catching this!

File: micro_dl/cli/image_inference.py [hunk 16]
```diff
@@ -122,10 +148,15 @@ def run_prediction(args, gpu_ids, gpu_mem_frac):
...
+    n_test_row = len(test_ids)
     for id in iter_ids:
         if id != split_idx_name:
-            metadata_ids[id] = np.unique(frames_meta[id])
-
+            metadata_ids[id] = np.unique(test_tile_meta[id])
+            n_test_row *= len(metadata_ids[id])
+    # create empty dataframe for test image metadata
+    test_frames_meta = pd.DataFrame(columns=
+                                    frames_meta.columns.values.tolist()+metrics,
```
[the remark this answers is not in the document]
Comment (smguo) [reply micro_dl/cli/image_inference.py]: You're right! I'll change this

File: micro_dl/cli/image_inference.py [hunk 17]
```diff
@@ -229,33 +261,64 @@ def run_prediction(args, gpu_ids, gpu_mem_frac):
...
+                # assuming target and predicted images are always 2D for now
+                # Load target
+                meta_idx = aux_utils.get_meta_idx(
+                    frames_meta,
+                    time_idx,
+                    target_channel,
+                    slice_idx,
+                    pos_idx,
+                )
+                file_path = os.path.join(
```
[the remark this answers is not in the document]
Comment (smguo) [reply micro_dl/cli/image_inference.py]: Thanks for catching this! Will remove it

File: micro_dl/plotting/plot_utils.py (hunk 12 above)
Comment (smguo) [reply micro_dl/plotting/plot_utils.py]: I saw that the preprocessing config switched to 'image_format', 'data_format' is still used in the network config. Here the input image format has to match the the network input format. Should we use 'image_format' for both preprocessing and network configs?

File: micro_dl/cli/image_inference.py (hunk 14 above)
Comment (smguo) [reply micro_dl/cli/image_inference.py]: Yes! That makes more sense

**Comment by jennyfolkesson** [comment]
@smguo is this PR ready for re-review? It looks like you have some merge conflicts though, let me know if you need help rebasing.

**Comment by smguo** [comment]
@jennyfolkesson I sorted out the merge conflicts. It's ready for re-review

File: micro_dl/cli/image_inference.py (hunk 4 above)
Comment (jennyfolkesson) [reply micro_dl/cli/image_inference.py]: They are not interchangeable, data_format refers to channel location and image_format refers do the order of the 3 dimensions (default zyx).
The training currently gets the image_format from the preprocessing config, so maybe we should add it to the network config too since that's the only one being read here...
https://github.com/czbiohub/microDL/blob/master/micro_dl/cli/train_script.py#L199
For now it can default to assuming image_format = zyx.

File: micro_dl/cli/image_inference.py (hunk 5 above)
Comment (jennyfolkesson) [reply micro_dl/cli/image_inference.py]: So I guess:
# TODO: Add image_format option to network config
im_target = np.transpose(im_target, [2, 0, 1])
if data_format == 'channels_first':
    im_target = im_target[np.newaxis, ...]
else:
    im_target = im_target[..., np.newaxis]

File: micro_dl/cli/image_inference.py [hunk 18]
```diff
@@ -229,33 +261,61 @@ def run_prediction(args, gpu_ids, gpu_mem_frac):
...
+                    im_target = np.transpose(im_target, [2, 0, 1])
+                # add the batch dimension
+                im_target = im_target[np.newaxis, ...]
+                if len(im_target.shape) < len(im_stack.shape):
+                    im_target = im_target[np.newaxis, ...]   
+                metric_vals = model.evaluate(x=im_stack, y=im_target)
+                for metric, metric_val in zip([loss]+metrics, metric_vals):
+                    test_frames_meta_row[metric] = metric_val
+
+                test_frames_meta = test_frames_meta.append(test_frames_meta_row, ignore_index=True)
```
Comment (jennyfolkesson) [inline micro_dl/cli/image_inference.py]: We try to avoid line lengths over 80, so can I ask you to:
```
test_frames_meta = test_frames_meta.append(
    test_frames_meta_row, 
    ignore_index=True,
)
```

File: micro_dl/cli/image_inference.py [hunk 19]
```diff
@@ -122,10 +148,14 @@ def run_prediction(args, gpu_ids, gpu_mem_frac):
...
     iter_ids = ['slice_idx', 'pos_idx', 'time_idx']
+    n_test_row = len(test_ids)
     for id in iter_ids:
         if id != split_idx_name:
-            metadata_ids[id] = np.unique(frames_meta[id])
-
+            metadata_ids[id] = np.unique(test_tile_meta[id])
+            n_test_row *= len(metadata_ids[id])
+    # create empty dataframe for test image metadata
+    test_frames_meta = pd.DataFrame(columns=
```
Comment (jennyfolkesson) [inline micro_dl/cli/image_inference.py]: I hate to be nitpicking but... PEP8 says it's better do write:
```
test_frames_meta = pd.DataFrame(
    columns=frames_meta.columns.values.tolist()+metrics,
)
```

File: micro_dl/cli/image_inference.py (hunk 18 above)
Comment (smguo) [reply micro_dl/cli/image_inference.py]: Sure. I'll keep the line short

**Comment by smguo** [comment]
Thanks @jennyfolkesson! I made the changes as you suggested
